go-controller/pkg/ovn/master.go

```diff
@@ -1491,3 +1497,226 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+
+	_, failed = gatewaysFailed.Load(node.Name)
+	if failed || gatewayChanged(oldNode, node) || nodeSubnetChanged(oldNode, node) || hostAddressesChanged(oldNode, node) {
+		err := oc.syncNodeGateway(node, nil)
+		if err != nil {
+			if !util.IsAnnotationNotSetError(err) {
+				return err
+			}
+			gatewaysFailed.Store(node.Name, true)
+		} else {
+			gatewaysFailed.Delete(node.Name)
+		}
+	}
+	return nil
+}
+
+func (oc *Controller) deleteNodeObj(node *kapi.Node) error {
+	var gatewaysFailed sync.Map
+	var mgmtPortFailed sync.Map
+	var addNodeFailed sync.Map
+	var nodeClusterRouterPortFailed sync.Map
+
+	klog.V(5).Infof("Deleting Node %q. Removing the node from "+
+		"various caches", node.Name)
+
+	nodeSubnets, _ := util.ParseNodeHostSubnetAnnotation(node)
+	if err := oc.deleteNode(node.Name, nodeSubnets); err != nil {
+		return err
+	}
+	oc.lsManager.DeleteNode(node.Name)
+	addNodeFailed.Delete(node.Name)
+	mgmtPortFailed.Delete(node.Name)
+	gatewaysFailed.Delete(node.Name)
+	nodeClusterRouterPortFailed.Delete(node.Name)
+	return nil
+}
+
+// iterateRetryNodes checks if any outstanding Nodes exists
+// then tries to re-add them if so
+// updateAll forces all nodes to be attempted to be retried regardless
+func (oc *Controller) iterateRetryNodes(updateAll bool) {
+	oc.retryNodeLock.Lock()
+	defer oc.retryNodeLock.Unlock()
+	now := time.Now()
+	for nodeName, nEntry := range oc.retryNodes {
+		entry := nEntry.(*retryObjEntry)
+		if entry.Ignore {
+			// neither addition nor deletion is being retried
+			continue
+		}
+		var node *kapi.Node
+		// check if we need to create
+		if entry.NewObj != nil {
+			n := entry.NewObj.(*kapi.Node)
+			kNode, err := oc.watchFactory.GetNode(n.Name)
+			if err != nil && errors.IsNotFound(err) {
+				klog.Infof("%s node not found in the informers cache, not going to retry node setup", nodeName)
+				entry.NewObj = nil
+			} else {
+				node = kNode
+			}
+		}
+
+		entry.BackoffSec = entry.BackoffSec * 2
+		if entry.BackoffSec > 60 {
+			entry.BackoffSec = 60
+		}
+		backoff := (entry.BackoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		nTimer := entry.TimeStamp.Add(backoff)
+		if updateAll || now.After(nTimer) {
+			klog.Infof("Node Retry: %s retry node setup", nodeName)
+
+			// check if we need to delete
+			if entry.OldObj != nil {
+				klog.Infof("Node Retry: Removing old Node for %s", nodeName)
+				node, ok := entry.OldObj.(*kapi.Node)
+				if ok {
+					if err := oc.deleteNodeObj(node); err != nil {
+						klog.Infof("Node Retry delete failed for %s, will try again later: %v",
+							nodeName, err)
+						entry.TimeStamp = time.Now()
+						continue
+					}
+					// successfully cleaned up old node, remove it from the retry cache
+					entry.OldObj = nil
+				}
+			}
+
+			// create new node if needed
+			if node != nil {
+				klog.Infof("Node Retry: Creating new node for %s", nodeName)
+				if err := oc.addNodeObj(node); err != nil {
```

Comment:
right it will Store new failures, but it wont remove failures if it succeeds. So you could have:
1. Add node fail: one or more of these maps is updated with a failure = true. InitRetry
2. Sometime later, retry happens: calls addNodeObj. Now how do we clear out the failures from the sync maps?

Reply:
during iterateRetryNodes before recreate again I can clean old sync maps

Reply:
As I see it, `iterateRetryNodes` [removes a node from the maps ](https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-1f90f4648dc7d68c42854549c4257ddcfc9bd9a5b6fc480c74e3a23989914bc3R1703-R1706) and `addNodeObj`, if it fails, [adds it to the maps](https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-1f90f4648dc7d68c42854549c4257ddcfc9bd9a5b6fc480c74e3a23989914bc3R1516-R1520).

Reply:
I dont think this will work. The retry action is always an "addNode", which doesn't take into consideration the previous failure could have been an update. I think the easiest way is to consolidate the add/update functions into one. If you want I can push a commit showing how it would look.